### ISSUE TITLE
Problems packaging .proselintrc

### ISSUE THREAD

**Issue by carlsmedstad** [amperser/proselint#1200]
Hey :wave:

I'm running into issues packaging this for Arch due to the system config file `.proselintrc`. Since the project migrated from `setuptools` to `poetry` I use [dephell](https://github.com/dephell/dephell) to convert it back to `setuptools` during installtion, as is recommended  by Arch community. It does seem like this conversion does not produce the correct `package_data` in `setup.py` (probably due to this not being supported) resulting in `.proselintrc` not being installed at all.

While I could patch the package and do a number of work-arounds, I wonder if we could get rid of `.proselintrc` altogether by hardcoding the system defaults directly in `proselint`. Having `.proselintrc` as a Python file or a part of `proselint/tools.py` would make packaging for distros less complex and it would not hurt `pip` distribution as far as I can tell.

Not distributing the default config in file format would make it a bit harder to make a custom config from scratch, but this could be rectified by a `--dump-config` flag which would print the current config. This is how the Clang tools works, for example.

**Comment by Nytelife26** [comment]
It is not necessary to use `dephell` to convert to `setup.py` format. Using `poetry build` will output a tarball and wheel - and the tar can be extracted which contains a `setup.py` already.

In any case, though, I'm not sure I like the idea of a hardcoded default configuration. Let me know if the release we're going to put out (0.11.3) fixes this, and if not, I'll look into it more.

**Comment by carlsmedstad** [comment]
> It is not necessary to use dephell to convert to setup.py format. Using poetry build will output a tarball and wheel - and the tar can be extracted which contains a setup.py already.

I think Arch has landed on using `dephell` due to not wanting to have a build dependency on `poetry`, since `setuptools` is used for practically all Arch Python packages. Additionally, the source package on PyPI which contains the generated `setup.py` does not contain the tests. On the other hand, `dephell` is no longer maintained, so I'm a bit conflicted on how to package `poetry` based projects in general.

> In any case, though, I'm not sure I like the idea of a hardcoded default configuration.

Isn't this already the case though? A config file among the source, distributed as a part of it, is sort of hardcoded. If someone uses it as a regular system-wide config file they'll corrupt the package so it's meant to be static as far as I can tell. Given that premise, migrating it to Python would be sensible in my mind.

> Let me know if the release we're going to put out (0.11.3) fixes this, and if not, I'll look into it more.

Will do!

**Comment by Nytelife26** [comment]
> On the other hand, `dephell` is no longer maintained, so I'm a bit conflicted on how to package `poetry` based projects in general.

If there are many of them, I'd suggest using `poetry`. If not, I'm not entirely sure - hasn't there been mention of a well maintained `dephell` fork? I thought I heard about one a while ago when I was looking through AUR packages.

> Given that premise, migrating it to Python would be sensible in my mind.

You raise a good point. It's actually not at all like the default system configuration file is in an accessible standard place, either. I'll discuss it with @suchow later, but I think embedding it into the distribution might be the wiser choice.

> Will do!

The release is pushed now, so feel free to try again.

**Comment by Nytelife26** [comment]
I can confirm everything is bundled correctly from PyPI and using `pip`, so hopefully it'll work in your packaging too.

**Comment by carlsmedstad** [comment]
> The release is pushed now, so feel free to try again.

Unfortunately no changes to my situation. I've temporarily resolved it by explicitly installing the file, so no rush:

```sh
  install -Dm644 proselint/.proselintrc "$pkgdir/usr/lib/python3.9/site-packages/proselint/.proselintrc"
```

> EDIT: Given that proselint breaks entirely when none of the default files are present, I'd call that a fault. On that basis, I'll work on removing the need for it.

By this you mean embedding the defaults in the code, right? Looked a bit at the code, and my suggestion would be to make a class for this to avoid the dynamism and potential bugs from passing around a dict. The class (or a static instance of it) would also be a natural place to store default values. This coupled with using `mypy` would be the path I would take. If you'd like I can make a PR.

**Comment by Nytelife26** [comment]
> Unfortunately no changes to my situation. I've temporarily resolved it by explicitly installing the file.

Interesting. How are you building proselint and installing it for the binary, may I ask? I have only tested this using the wheel - I'm not sure if it also works for the tarball.

> By this you mean embedding the defaults in the code, right? 

Indeed I do.

> my suggestion would be to make a class for this to avoid the [...] potential bugs from passing around a dict.

That's what I was thinking, too. Something similar to how configuration is typically handled in Rust, wherein a structure is made and has default values and then configuration files are merged in which override the defaults.

> This coupled with using `mypy`

I use `pyright` locally as it is much more complete and flexible than `mypy` - I hadn't thought much about including it in CI, as Python tends to not be type-checked much if at all (given its nature as a dynamically typed language) and many people are not used to it, but I can open an RFC issue for that.

**Comment by carlsmedstad** [comment]
> Interesting. How are you building proselint and installing it for the binary, may I ask? I have only tested this using the wheel - I'm not sure if it also works for the tarball.

You can read the `PKGBUILD` here: https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=proselint

> but I can open an RFC issue for that.

Great RFC, let's have that discussion there and ignore that in this issue.

Would you like to work on this? If so could you assign yourself to the issue? If not, I'd happily do it :)

**Comment by Nytelife26** [comment]
Unfortunately `pyproject.toml` manifests utterly refuse to bundle dotfiles. Therefore, your solution is the only one.

Furthermore, I may open an RFC to change `.proselintrc` to `.proselintrc.json` so it has the standard file extension.

**Comment by mavit** [comment]
>  Additionally, the source package on PyPI which contains the generated `setup.py` does not contain the tests.

The tests are back since #1195.

**Comment by mavit** [comment]
Removing the need for a default `.proselintrc` sounds like a longer-term thing, so in the short term can I propose again that we rename it to `proselintrc` as in b6d8eb6e98398d0b5e6ce965ad95b4efd7f06889?

**Comment by Nytelife26** [comment]
> Removing the need for a default `.proselintrc` sounds like a longer-term thing, so in the short term can I propose again that we rename it to `proselintrc`

Making any changes to its location or name would require approximately the same amount of effort. For that reason, I suggest we make this atomic and avoid doing it except for when absolutely necessary. Therefore, unless there are objections, I'll likely be removing the need for a default in the next version (alongside the migration from `.proselintrc` to `.proselintrc.json`)

**Comment by carlsmedstad** [comment]
@Nytelife26 Sounds good, I can review your PR about the removal if you'd like.

**Comment by Nytelife26** [comment]
Okay, so here are the solutions so far:

- [ ] Build a default config into Proselint (accomplished)
- [ ] Refactor `.proselintrc`s to use `.proselintrc.json` (accomplished)
- [ ] Add CLI flag to dump default config
- [ ] Refactor `.proselintrc.json` format to use `include` and `exclude` directives

**Comment by Nytelife26** [comment]
This should be fixed in the next release. Thank you all.

I'll reopen this for now so we can report back once the next release has been made.

**Comment by Nytelife26** [comment]
@carlsmedstad Does this issue persist in 0.12.0?

**Comment by carlsmedstad** [comment]
It does not. Thank you for addressing it!
